### PR TITLE
Optimize VRAM use in textual inversion training

### PR DESCRIPTION
**Pull request by Ttl**
Cast frozen modules to fp16/bf16 when using mixed precision. Add gradient checkpoint command line option.

OOMs before on my 8 GB VRAM GPU. With these changes and using `--mixed_precision=fp16 --gradient_checkpointing` VRAM use is 6341 MB and the results look good.

### CODE REVIEW

File: examples/textual_inversion/textual_inversion.py [hunk 1]
```diff
@@ -511,10 +526,11 @@ def main():
     for epoch in range(args.num_train_epochs):
         text_encoder.train()
         for step, batch in enumerate(train_dataloader):
-            with accelerator.accumulate(text_encoder):
+            with accelerator.autocast(), accelerator.accumulate(text_encoder):
```
Comment (keturn) [inline examples/textual_inversion/textual_inversion.py:529]: What would be necessary in order for this to work without autocast?

There's some concern about its use: https://github.com/huggingface/diffusers/pull/511

**Comment by keturn** [comment]
@isamu-isozaki is this the approach you've been using?

**Comment by isamu-isozaki** [comment]
@keturn Pretty much! I didn't know about the autocast functionality so I manually moved most of the parts to cpu and cuda. The code is [here](https://github.com/isamu-isozaki/diffusers). One thing I remember was that for 6gb of ram, it'll omm before that accelerator.accumulate part so most models are better moved to the cpu.

File: examples/textual_inversion/textual_inversion.py (hunk 1 above)
Comment (Ttl) [reply examples/textual_inversion/textual_inversion.py:529]: Speed is about 10% better without autocast and VRAM use decreases to 5500 MB. It does need some additional casts and I don't think it's quite identical functionally since some operations that would be autocasted to fp32 are computed in fp16. Results look still fine though so it might be okay. Smaller VRAM use does enable increasing batch size or disabling gradient checkpointing for even bigger speed up.

**Comment by isamu-isozaki** [comment]
Hi! Great pr! @Ttl @keturn. It doesn't fit in 6gb ram as it is now but once I did this
```
slice_size = unet.config.attention_head_dim // 2
unet.set_attention_slice(slice_size)
```
it fits. Thanks for this! Now my training will get way better.

**Comment by isamu-isozaki** [comment]
@keturn sry on second thought this is way different from my approach but it's way better too!

**Comment by keturn** [comment]
If you add <code>revision="[fp16](https://huggingface.co/CompVis/stable-diffusion-v1-4/tree/fp16)"</code> to `from_pretrained`, do you still have to do the conversions to `weight_dtype`?

File: examples/textual_inversion/textual_inversion.py (hunk 1 above)
Comment (isamu-isozaki) [reply examples/textual_inversion/textual_inversion.py:529]: I checked a bit. The main problem is 
```
accelerator.backward(loss)
```
If that can be done in fp16, it should work without autocast

**Comment by patrickvonplaten** [comment]
@patil-suraj can you take a look here?

**Comment by Ttl** [comment]
I'm using locally saved weights and adding `revision="fp16"` doesn't seem to do anything in that case.

**Comment by isamu-isozaki** [comment]
I tried revision fp16 and got oom for some reason. will double check later today

**Review by patil-suraj — COMMENTED** [review]
Thanks a lot for the PR! As explained in the comments, we shouldn't cast all weights to half-precision unless specified by user and `autocast` should not be used as default as it won't allow full-precision training.

Also, please note that, the examples scripts are just examples, they show how to do a certain example with simple and easy way. For more customization it is recommended that the user should modify the script on their own as they need it. This will help keep the script simple so any one can understand and modify it for themselves if needed.

if you just pass `mixed_precision="fp16`, `accelerate` should enable mixed-precision without any code changes.


I'm not in favor of this, Hope you understand, thanks a lot!

**Comment by Ttl** [comment]
The idea with casting the weights of non-trained nets is that without it fp32 weights are transferred to vram even when training in fp16. Since they are not trained we don't need to keep fp32 copy of them in vram.

`self.training` is controlled by `train()` or `eval()` call of the module. Since in this case we have set unet to be in `eval()` without removing the `self.training` gradient checkpointing is not enabled. Gradient checkpointing is useful in this case since we need to store activations in the unet for backwards pass since it's between our trainable weights and loss calculation. I checked that enabling it saves 1080 MB of memory.

I can maintain a copy of the script that casts non-trained weights to fp16 locally, but it would be nice if the gradient checkpointing changes would be merged. Would you be fine with that?

**Comment by patil-suraj** [comment]
> Gradient checkpointing is useful in this case since we need to store activations in the unet for backwards pass since it's between our trainable weights and loss calculation. I checked that enabling it saves 1080 MB of memory.

That's a really good observation! Sorry, I rushed the review a bit. In this case keeping the gradient checkpointing changes makes sense, let me try it quickly and get back to you.

Thanks a lot!

**Comment by patil-suraj** [comment]
Also pinging @patrickvonplaten and @anton-l . Are the activations stored even when the grads are disabled for the model ?

**Comment by Ttl** [comment]
I added a commit that removes the autocast. It should work with fp32 and bf16 too but I can't test it on my GPU. This PR does have a side effect that it saves fp16 quantized weights of unet and vae since fp32 weights from those were discarded if training in fp16. If you prefer I can remove the training changes and only keep the gradient checkpointing change.

**Comment by patrickvonplaten** [comment]
I think this PR is currently blocked by:
- `gradient_checkpointing` being a bit flaky when model is set to `train` mode
- that we're not able to pass dropout down

cc @patil-suraj

**Comment by Thomas-MMJ** [comment]
Any progress on the blockers?

**Comment by patrickvonplaten** [comment]
@patil-suraj could you have another look here?

**Review by patil-suraj — COMMENTED** [review]
Sorry to only comeback to this now!

- the output of `gradient_checkpointing` enabled model in `train` mode is deterministic when dropout is zero (which is the default), I ran a few tests to confirm this.
- We have dropout set to 0 by default, and need to allow passing it to the model. But this could be added later.

So there is no blocker for this PR now. @Ttl could you please adapt the PR to put the `unet` in `train` mode, and not modify the gradient checkpointing parts. Then I'll open a follow-up PR to allow passing `dropout` to the model that we can set it 0 to always make sure we have deterministic output from `unet`.

**Comment by patrickvonplaten** [comment]
cc @patil-suraj here - could you maybe go into the PR if author doesn't reply anymore to not forget it?

**Comment by patrickvonplaten** [comment]
Taking over this PR to remove the deprecation warning

**Comment by patrickvonplaten** [comment]
cc @patil-suraj - could you maybe post some instructions here on how to proceed? Then someone else could pick it up

**Comment by patil-suraj** [comment]
Sorry for being late again, I've posted instructions in this comment https://github.com/huggingface/diffusers/pull/687#pullrequestreview-1165354545

@Ttl LMK if you are busy, then I'll make the necessary changes and merge :)

**Comment by Ttl** [comment]
It's been quite long since I last looked at this code and I haven't used textual inversion much anymore. Feel free to make necessary changes to get it merged if you want to.

**Comment by patil-suraj** [comment]
Thanks, will make open a PR then :)
